02_Util/src/security/SignedMeterValuesUtil.ts

```diff
@@ -0,0 +1,225 @@
+  }
+
+  /**
+   * Checks the validity of a meter value.
+   *
+   * If a meter value is unsigned, it is valid.
+   *
+   * If a meter value is signed, it is valid if:
+   * - SignedMeterValuesConfig is configured
+   * AND
+   * - The incoming signed meter value's signing method matches the configured signing method
+   * AND
+   * - The incoming signed meter value's public key is empty but there is a public key stored for that charging station
+   * OR
+   * - The incoming signed meter value's public key isn't empty and it matches the configured public key
+   *
+   * @param stationId - The charging station the meter values belong to
+   * @param meterValues - The list of meter values
+   */
+  public async validateMeterValues(
+    stationId: string,
+    meterValues: [MeterValueType, ...MeterValueType[]],
+  ): Promise<boolean> {
+    let validMeterValues = true;
+
+    const publicKeyFrequencyVariableAttribute =
+      await this._deviceModelRepository.readAllByQuerystring({
+        stationId,
+        component_name: 'OCPPCommCtrlr',
+        variable_name: 'PublicKeyWithSignedMeterValue',
+      });
+
+    const shouldPublicKeyBeUsedForValidation =
+      publicKeyFrequencyVariableAttribute.length > 0 &&
+      publicKeyFrequencyVariableAttribute[0].value !== 'Never';
+
+    if (shouldPublicKeyBeUsedForValidation) {
+      for (const meterValue of meterValues) {
+        for (const sampledValue of meterValue.sampledValue) {
+          validMeterValues =
+            validMeterValues &&
+            (await this.isSignedSampledValueValid(stationId, sampledValue));
+        }
+      }
+    }
+
+    return validMeterValues;
+  }
+
+  private async isMeterValueSignatureValid(
+    signedMeterValue: SignedMeterValueType,
+    publicKeyFileId?: string,
+  ): Promise<boolean> {
+    const incomingPublicKeyString = signedMeterValue.publicKey;
+    const signingMethod = signedMeterValue.signingMethod;
+
+    if (
+      !this._signedMeterValuesConfiguration?.publicKeyFileId ||
+      (publicKeyFileId &&
+        publicKeyFileId !==
+          this._signedMeterValuesConfiguration?.publicKeyFileId) ||
+      (!publicKeyFileId && incomingPublicKeyString.length === 0) ||
+      this._signedMeterValuesConfiguration?.signingMethod !== signingMethod
+    ) {
+      return false;
+    }
+
+    const configuredPublicKey = this.formatKey(
+      (
+        await this._fileAccess.getFile(
+          this._signedMeterValuesConfiguration.publicKeyFileId,
+        )
+      ).toString(),
+    );
+
+    if (incomingPublicKeyString.length > 0) {
+      const signedMeterValuePublicKey = Buffer.from(
+        signedMeterValue.publicKey,
+        'base64',
+      ).toString();
+
+      if (configuredPublicKey !== signedMeterValuePublicKey) {
+        return false;
+      }
+    }
+
+    switch (signingMethod) {
+      case 'RSASSA-PKCS1-v1_5':
+        try {
+          const cryptoPublicKey = await crypto.subtle.importKey(
```

Comment:
Could dump the contents of this case into its own `validateRSASignature(...)` method

Reply:
moved into its own helper method 👍

Reply:
@yuvenus I don't think that change got through?